File: 1679-shortest-subarray-to-be-removed-to-make-array-sorted/shortest-subarray-to-be-removed-to-make-array-sorted.cpp
class Solution {
public:
    int findLengthOfShortestSubarray(vector<int>& arr) {
        int n = arr.size();
        int r = n - 1;
        while (r > 0 && arr[r - 1] <= arr[r])
            r -= 1;
        int res = r;

        int l = 0;
        while (l < n - 1 && arr[l] <= arr[l + 1])
            l+= 1;
        res = min(res, n - l - 1);

        l = 0;
        r = n - 1;
        while (l < r){
            while (r < n && l + 1 < r && arr[r - 1] <= arr[r] && arr[l] <= arr[r])
                r -= 1;
            
            while (r < n && arr[l] > arr[r])
                r += 1;
            
            res = min(res, r - l - 1);
            if (arr[l] > arr[l + 1]) break;

            l += 1;
        }
        return res;
    }
};

File: 1679-shortest-subarray-to-be-removed-to-make-array-sorted/shortest-subarray-to-be-removed-to-make-array-sorted.py
class Solution:
    def findLengthOfShortestSubarray(self, arr: List[int]) -> int:
        n = len(arr) - 1
        r = n
        while r > 0 and arr[r - 1] <= arr[r]:
            r -= 1
        res = r
        
        l = 0
        while l < n  and arr[l] <= arr[l + 1]:
            l += 1
        
        res = min(res, n - l)

        l = 0
        r = n

        while l < r:
            while r <= n and l + 1 < r and arr[r - 1] <= arr[r] and arr[l] <= arr[r]:
                r -= 1
            
            while r <= n  and arr[l] > arr[r]:
                r += 1

            res = min(res, r - l - 1)

            if arr[l] > arr[l + 1]:
                break

            l += 1
            
            
        return res